FBRetainCycleDetector/Layout/Classes/Parser/FBStructEncodingParser.mm

```diff
@@ -9,95 +9,152 @@
     }
-  }
-
-  NSPredicate *filterPredicate = [NSPredicate predicateWithBlock:^BOOL(id evaluatedObject,
-                                                                       NSDictionary<NSString *, id> *bindings) {
-    if ([evaluatedObject isKindOfClass:[FBParsedType class]] ||
-        [evaluatedObject isKindOfClass:[FBParsedStruct class]]) {
-      return YES;
+    
+    char currentCharacter() {
+      return string[index];
+    }
+    
+    std::string scanUpToCharacterFromSet(std::string &characterSet) {
+      size_t pos = string.find_first_of(characterSet, index);
+      if (pos == std::string::npos) {
+        index = string.length();
+        return "";
+      }
+      
+      std::string inBetweenString = string.substr(index, pos-index);
+      index = pos;
+      return inBetweenString;
     }
-    return NO;
-  }];
-  NSArray<FBParsedType *> *finalTypes = [types filteredArrayUsingPredicate:filterPredicate];
-
-  return {
-    .containedTypes = finalTypes,
-    .typeName = structTypeName,
   };
-}
-
-FBParsedStruct *FBParseStructEncoding(NSString *structEncodingString) {
-  return FBParseStructEncodingWithName(structEncodingString, nil);
-}
+  
+};
 
-FBParsedStruct *FBParseStructEncodingWithName(NSString *structEncodingString, NSString *structName) {
-  struct _StructParseResult result = _ParseStructEncodingWithScanner([NSScanner scannerWithString:structEncodingString]);
-  FBParsedStruct *outerStruct = [[FBParsedStruct alloc] initWithName:structName
-                                                        typeEncoding:structEncodingString
-                                                      structTypeName:result.typeName
-                                              typesContainedInStruct:result.containedTypes];
-  [outerStruct passTypePath:nil];
-  return outerStruct;
-}
+namespace FB { namespace RetainCycleDetector { namespace Parser {
+  
+  /**
+   Intermediate struct object used inside the algorithm to pass some
+   information when parsing nested structures.
+   */
+  struct _StructParseResult {
+    std::vector<std::shared_ptr<Type>> containedTypes;
+    std::string typeName;
+  };
+  
+  static struct _StructParseResult _ParseStructEncodingWithScanner(_StringScanner &scanner) {
+    std::vector<std::shared_ptr<BaseType>> types;
+    
+    // Every struct starts with '{'
+    __unused bool scannedCorrectly = scanner.scanString("{");
+    NSCAssert(scannedCorrectly, @"The first character of struct encoding should be {");
+    
+    // Parse name
+    std::string structTypeName = scanner.scanUpToString("=");
+    scanner.scanString("=");
+    
+    std::string literalEndingCharacters = "\"}";
+    
+    while (!(scanner.scanString("}"))) {
+      if (scanner.scanString("\"")) {
+        std::string parseResult = scanner.scanUpToString("\"");
+        scanner.scanString("\"");
+        if (parseResult.length() > 0) {
+          types.emplace_back(std::shared_ptr<Unresolved>(new Unresolved(parseResult)));
```

Comment:
Two notes on this line:

1. `emplace_back` is being called with an already constructed shared pointer, so should probably be `push_back`. Otherwise, this would be doing an extra (needless) copy via `shared_ptr`'s copy constructor.
2. Instead of using a `shared_ptr` constructor, `std::make_shared` is a recommended way to construct instances.